segment/writer.go

```diff
@@ -367,15 +367,46 @@ func (w *Writer) appendEntry(e types.LogEntry) error {
 		typ: FrameEntry,
 		len: uint32(len(e.Data)),
 	}
-	return w.appendFrame(fh, e.Data)
+
+	bufOffset, err := w.appendFrame(fh, e.Data)
+	if err != nil {
+		return err
+	}
+
+	// updating the index
+	offsets := w.getOffsets()
+
+	// Add the index entry. Note this is safe despite mutating the same backing
+	// array as tail because it's beyond the limit current readers will access
+	// until we do the atomic update below. Even if append re-allocates the
+	// backing array, it will only read the indexes smaller than numEntries from
+	// the old array to copy them into the new one and we are not mutating the
+	// same memory locations. Old readers might still be looking at the old
+	// array (lower than numEntries) through the current tail.offsets slice but
+	// we are not touching that at least below numEntries.
+	if len(offsets) == 0 && w.info.BaseIndex == 1 {
+		for i := 0; i < int(e.Index)-1; i++ {
+			offsets = append(offsets, 0)
+		}
+
+		atomic.StoreUint64(&w.info.MinIndex, e.Index)
```

Comment:
I see this is an atomic store now, but I don't see atomic Loads for the places that read this (e.g. OffsetForFrame above on line 356).

Also didn't see the update to the concurrent test in `writer_test.go` that would help confirm when the racey behaviour is fixed.

Finally, because we are just updating this in place in a struct that is used for other purposes, there is a chance that in the future this struct could be modified so this field is no longer 64 bit aligns (it does happen to be right now but there's nothing to stop the struct being changed in the future since the requirement is undocumented). I think we either should:
 1. Store a separate field in the `writer {}` state for this and use it here and in `OffsetForFrame` above. We'd probably want to document why it might be different to `info.MinIndex`. Downside is that its' not obvious that info.MinIndex might not be correct any more.
 2. Update the docs on the `SegmentInfo` struct fields to point out that we use them internally and rely on 64 bit alignment so not to change that. This is kinda gross as they are in a separate package, but at least reduces the risk of introducing a breaking change later accidentally.
 3. Make the whole `info` field an `atomic.Value` and copy-and-swap it out here. This is probably the cleanest overall?

Reply:
please check https://github.com/hashicorp/raft-wal/pull/23/commits/483142e708fab49a99fc7ccc936584e3a53473eb

Reply:
I will work on the test next!